File: framework/languages/node/samples/Bingo.Ts/Shared/Contracts/messages.ts
const PacketNames = Object.freeze({
  authenticateReq: 'AuthenticateReq',
  authenticatePlayerReq: 'AuthenticatePlayerReq',
  ensurePlayerActorReq: 'EnsurePlayerActorReq',
  matchBingoReq: 'MatchBingoReq',
  matchBingoApiReq: 'MatchBingoApiReq',
  allocateBingoRoom: 'AllocateBingoRoomReq',
  submitBingoCardReq: 'SubmitBingoCardReq',
  observeBingoEventsReq: 'ObserveBingoEventsReq',
  stopObservingBingoEventsReq: 'StopObservingBingoEventsReq',
  bingoRewardAcquiredEvent: 'BingoRewardAcquiredMsg',
  pingMsg: 'PingMsg',
  playerJoinedNotify: 'PlayerJoinedNotify',
  gameStartedNotify: 'BingoGameStartedNotify',
  numberDrawnNotify: 'BingoNumberDrawnNotify',
  stateNotify: 'BingoStateNotify',
  gameEndedNotify: 'BingoGameEndedNotify',
  rewardAnnouncedNotify: 'BingoRewardAnnouncedNotify'
});

const BingoModes = Object.freeze({
  twoPlayer: 'two-player'
});

const BingoSamplePlayers = Object.freeze({
  player1: 'player-1',
  player2: 'player-2',
  observer: 'observer'
});

const BingoRewardItems = Object.freeze({
  goldenDauberId: 'golden-dauber',
  goldenDauberName: 'Golden Dauber',
  legendaryRarity: 'Legendary'
});

export type BingoMode = typeof BingoModes.twoPlayer;

export enum BingoRoomStatus {
  WaitingForPlayers = 'WaitingForPlayers',
  Running = 'Running',
  Finished = 'Finished'
}

export class AuthenticateReq {
  accessToken: string;

  constructor(accessToken: string) {
    this.accessToken = accessToken;
  }
}

export class AuthenticatePlayerReq {
  accessToken: string;

  constructor(accessToken: string) {
    this.accessToken = accessToken;
  }
}

export interface AuthenticatePlayerRes {
  accepted: boolean;
  actorId: string | null;
  displayName: string | null;
  reason: string | null;
}

export interface AuthenticateSessionRes {
  actorId: string;
  displayName: string;
  actorNodeRid: string;
}

export class MatchBingoReq {
  mode?: BingoMode;
  actorId?: string;
  displayName?: string;

  constructor(mode?: BingoMode) {
    if (mode !== undefined) {
      this.mode = mode;
    }
  }
}

export class MatchBingoApiReq {
  actorId: string;
  displayName: string;
  mode: BingoMode;
  actorNodeRid: string;

  constructor(actorId: string, displayName: string, mode: BingoMode, actorNodeRid: string) {
    this.actorId = actorId;
    this.displayName = displayName;
    this.mode = mode;
    this.actorNodeRid = actorNodeRid;
  }
}

export interface MatchBingoApiRes {
  roomId: string;
  roomOwnerNodeRid: string;
}

export class AllocateBingoRoomReq {
  mode: BingoMode;

  constructor(mode: BingoMode) {
    this.mode = mode;
  }
}

export interface AllocateBingoRoomRes {
  roomId: string;
  roomOwnerNodeRid: string;
}

export interface BingoRoomSettingsPayload {
  roomName: string;
  mode: BingoMode;
  requiredPlayers: number;
  maxDrawNumber: number;
  purpose: string;
  observedRoomId: string | null;
}

export class EnsurePlayerActorReq {
  actorId: string;
  displayName: string;

  constructor(actorId: string, displayName: string) {
    this.actorId = actorId;
    this.displayName = displayName;
  }
}

export interface BingoActorRef {
  nodeRid: string;
  actorId: string;
  generation: number;
}

export interface EnsurePlayerActorRes {
  actorId: string;
  actorType: string;
  actor: BingoActorRef;
}

export interface BingoRoomJoinReq {
  roomId: string;
  actorId: string;
  displayName: string;
  observeOnly: boolean;
}

export interface BingoRoomJoinRes {
  state: unknown;
}

export interface MatchBingoRes {
  roomId: string;
  state: unknown;
  roomOwnerNodeRid: string;
}

export class SubmitBingoCardReq {
  roomId: string;
  card: number[];
  actorId?: string;
  displayName?: string;

  constructor(roomId: string, card: number[]) {
    this.roomId = roomId;
    this.card = card;
  }
}

export interface SubmitBingoCardRes {
  state: unknown;
}

export class ObserveBingoEventsReq {
  roomId: string;

  constructor(roomId: string) {
    this.roomId = roomId;
  }
}

export interface ObserveBingoEventsRes {
  subscribed: boolean;
  observerNodeRid: string;
}

export class StopObservingBingoEventsReq {
  roomId: string;

  constructor(roomId: string) {
    this.roomId = roomId;
  }
}

export interface StopObservingBingoEventsRes {
  stopped: boolean;
  observerNodeRid: string;
}

export interface RejectedCommandRes {
  rejected: true;
  reason: string;
}

export interface PlayerIdentity {
  actorId: string;
  displayName: string;
}

export class PlayerJoinedNotify {
  constructor(
    readonly roomId: string,
    readonly actorId: string,
    readonly displayName: string,
    readonly seat: number,
    readonly isHost: boolean,
    readonly state: unknown
  ) {}
}

export class BingoGameStartedNotify {
  constructor(readonly state: unknown) {}
}

export class BingoStateNotify {
  constructor(readonly state: unknown) {}
}

export class BingoGameEndedNotify {
  constructor(readonly state: unknown) {}
}

export type StateEnvelope = BingoGameStartedNotify | BingoStateNotify | BingoGameEndedNotify;

export interface RoomJoinError {
  error: string;
}

export class BingoNumberDrawnNotify {
  constructor(
    readonly roomId: string,
    readonly drawSeq: number,
    readonly number: number,
    readonly state: unknown
  ) {}
}

export type NumberDrawnNotify = BingoNumberDrawnNotify;

export class BingoRewardAcquiredMsg {
  constructor(
    readonly roomId: string,
    readonly actorId: string,
    readonly drawSeq: number,
    readonly itemId: string,
    readonly itemName: string,
    readonly rarity: string
  ) {}
}

export class BingoRewardAnnouncedNotify implements BingoRewardAcquiredMsg {
  constructor(
    readonly roomId: string,
    readonly actorId: string,
    readonly drawSeq: number,
    readonly itemId: string,
    readonly itemName: string,
    readonly rarity: string,
    readonly receivingSpotNodeRid: string
  ) {}
}

function actorDisplayName(actorId: string): string {
  return actorId.replace('player-', 'Player ');
}

function deterministicCard(actorId: string): number[] {
  return {
    [BingoSamplePlayers.player1]: [1, 2, 3, 4, 0, 6, 7, 8, 9],
    [BingoSamplePlayers.player2]: [10, 11, 12, 13, 0, 14, 15, 1, 2]
  }[actorId] ?? [1, 3, 5, 7, 0, 9, 11, 13, 15];
}

function authenticateReq(accessToken: string): AuthenticateReq {
  return new AuthenticateReq(accessToken);
}

function authenticatePlayerReq(accessToken: string): AuthenticatePlayerReq {
  return new AuthenticatePlayerReq(accessToken);
}

function authenticatePlayerAccepted(accessToken: string): AuthenticatePlayerRes {
  return {
    accepted: true,
    actorId: accessToken,
    displayName: actorDisplayName(accessToken),
    reason: null
  };
}

function authenticatePlayerRejected(reason: string): AuthenticatePlayerRes {
  return {
    accepted: false,
    actorId: null,
    displayName: null,
    reason
  };
}

function authenticateSessionRes(actorId: string, displayName: string, actorNodeRid: string): AuthenticateSessionRes {
  return { actorId, displayName, actorNodeRid };
}

function matchBingoReq(mode: BingoMode = BingoModes.twoPlayer): MatchBingoReq {
  return new MatchBingoReq(mode);
}

function matchBingoApiReq(
  actorId: string,
  displayName: string,
  mode: BingoMode | undefined,
  actorNodeRid: string
): MatchBingoApiReq {
  return new MatchBingoApiReq(actorId, displayName, mode ?? BingoModes.twoPlayer, actorNodeRid);
}

function matchBingoApiRes(roomId: string, roomOwnerNodeRid: string): MatchBingoApiRes {
  return { roomId, roomOwnerNodeRid };
}

function allocateBingoRoomReq(mode: BingoMode = BingoModes.twoPlayer): AllocateBingoRoomReq {
  return new AllocateBingoRoomReq(mode);
}

function allocateBingoRoomRes(roomId: string, roomOwnerNodeRid: string): AllocateBingoRoomRes {
  return { roomId, roomOwnerNodeRid };
}

function bingoRoomSettingsPayload(settings: {
  roomName: string;
  mode: BingoMode;
  requiredPlayers: number;
  maxDrawNumber: number;
  purpose?: string;
  observedRoomId?: string | null;
}): BingoRoomSettingsPayload {
  return {
    roomName: settings.roomName,
    mode: settings.mode,
    requiredPlayers: settings.requiredPlayers,
    maxDrawNumber: settings.maxDrawNumber,
    purpose: settings.purpose ?? 'Game',
    observedRoomId: settings.observedRoomId ?? null
  };
}

function ensurePlayerActorReq(actorId: string, displayName: string): EnsurePlayerActorReq {
  return new EnsurePlayerActorReq(actorId, displayName);
}

function ensurePlayerActorRes(actor: BingoActorRef): EnsurePlayerActorRes {
  return {
    actorId: actor.actorId,
    actorType: 'bingo.player',
    actor
  };
}

function bingoRoomJoinReq(roomId: string, actorId: string, displayName: string, observeOnly = false): BingoRoomJoinReq {
  return { roomId, actorId, displayName, observeOnly };
}

function matchBingoRes(roomId: string, state: unknown, roomOwnerNodeRid: string): MatchBingoRes {
  return { roomId, state, roomOwnerNodeRid };
}

function submitBingoCardReq(roomId: string, card: number[]): SubmitBingoCardReq {
  return new SubmitBingoCardReq(roomId, card);
}

function submitBingoCardRes(state: unknown): SubmitBingoCardRes {
  return { state };
}

function observeBingoEventsReq(roomId: string): ObserveBingoEventsReq {
  return new ObserveBingoEventsReq(roomId);
}

function observeBingoEventsRes(subscribed: boolean, observerNodeRid: string): ObserveBingoEventsRes {
  return { subscribed, observerNodeRid };
}

function stopObservingBingoEventsReq(roomId: string): StopObservingBingoEventsReq {
  return new StopObservingBingoEventsReq(roomId);
}

function stopObservingBingoEventsRes(stopped: boolean, observerNodeRid: string): StopObservingBingoEventsRes {
  return { stopped, observerNodeRid };
}

function rejectedCommandRes(reason: string): RejectedCommandRes {
  return { rejected: true, reason };
}

function withPlayerIdentity<TRequest extends object>(
  request: TRequest,
  actorId: string,
  displayName: string
): TRequest & PlayerIdentity {
  const value = request as TRequest & Partial<PlayerIdentity>;
  value.actorId = actorId;
  value.displayName = displayName;
  return value as TRequest & PlayerIdentity;
}

function playerJoinedNotify(
  roomId: string,
  actor: { actorId: string; displayName: string },
  seat: number,
  isHost: boolean,
  state: unknown
): PlayerJoinedNotify {
  return new PlayerJoinedNotify(roomId, actor.actorId, actor.displayName, seat, isHost, state);
}

function roomJoinError(error: string): RoomJoinError {
  return { error };
}

function numberDrawnNotify(roomId: string, drawSeq: number, number: number, state: unknown): BingoNumberDrawnNotify {
  return new BingoNumberDrawnNotify(roomId, drawSeq, number, state);
}

function bingoRewardAcquiredEvent(value: BingoRewardAcquiredMsg): BingoRewardAcquiredMsg {
  return new BingoRewardAcquiredMsg(
    value.roomId,
    value.actorId,
    value.drawSeq,
    value.itemId,
    value.itemName,
    value.rarity
  );
}

function bingoRewardAnnouncedNotify(
  event: BingoRewardAcquiredMsg,
  receivingSpotNodeRid: string
): BingoRewardAnnouncedNotify {
  return new BingoRewardAnnouncedNotify(
    event.roomId,
    event.actorId,
    event.drawSeq,
    event.itemId,
    event.itemName,
    event.rarity,
    receivingSpotNodeRid
  );
}

export {
  BingoModes,
  BingoRewardItems,
  BingoSamplePlayers,
  PacketNames,
  actorDisplayName,
  allocateBingoRoomReq,
  allocateBingoRoomRes,
  authenticatePlayerAccepted,
  authenticatePlayerRejected,
  authenticatePlayerReq,
  authenticateReq,
  authenticateSessionRes,
  bingoRewardAcquiredEvent,
  bingoRewardAnnouncedNotify,
  bingoRoomJoinReq,
  bingoRoomSettingsPayload,
  deterministicCard,
  ensurePlayerActorReq,
  ensurePlayerActorRes,
  matchBingoApiRes,
  matchBingoReq,
  matchBingoApiReq,
  matchBingoRes,
  numberDrawnNotify,
  observeBingoEventsReq,
  observeBingoEventsRes,
  playerJoinedNotify,
  rejectedCommandRes,
  roomJoinError,
  stopObservingBingoEventsReq,
  stopObservingBingoEventsRes,
  submitBingoCardReq,
  submitBingoCardRes,
  withPlayerIdentity
};

File: framework/languages/java/samples/kotlin/Bingo/Shared/src/main/kotlin/systems/zlink/samples/kotlin/bingo/shared/contracts/Messages.kt
package systems.zlink.samples.kotlin.bingo.shared.contracts

typealias AuthenticateReq = Messages.AuthenticateReq
typealias AuthenticateRes = Messages.AuthenticateRes
typealias AuthenticatePlayerReq = Messages.AuthenticatePlayerReq
typealias AuthenticatePlayerRes = Messages.AuthenticatePlayerRes
typealias EnsurePlayerActorReq = Messages.EnsurePlayerActorReq
typealias ActorRefWire = Messages.ActorRefWire
typealias EnsurePlayerActorRes = Messages.EnsurePlayerActorRes
typealias MatchBingoReq = Messages.MatchBingoReq
typealias MatchBingoRes = Messages.MatchBingoRes
typealias MatchBingoApiReq = Messages.MatchBingoApiReq
typealias MatchBingoApiRes = Messages.MatchBingoApiRes
typealias AllocateBingoRoomReq = Messages.AllocateBingoRoomReq
typealias AllocateBingoRoomRes = Messages.AllocateBingoRoomRes
typealias BingoRoomSettingsPayload = Messages.BingoRoomSettingsPayload
typealias BingoRoomJoinReq = Messages.BingoRoomJoinReq
typealias BingoRoomJoinRes = Messages.BingoRoomJoinRes
typealias SubmitBingoCardReq = Messages.SubmitBingoCardReq
typealias SubmitBingoCardRes = Messages.SubmitBingoCardRes
typealias ObserveBingoEventsReq = Messages.ObserveBingoEventsReq
typealias ObserveBingoEventsRes = Messages.ObserveBingoEventsRes
typealias StopObservingBingoEventsReq = Messages.StopObservingBingoEventsReq
typealias StopObservingBingoEventsRes = Messages.StopObservingBingoEventsRes
typealias PlayerJoinedNotify = Messages.PlayerJoinedNotify
typealias BingoGameStartedNotify = Messages.BingoGameStartedNotify
typealias BingoNumberDrawnNotify = Messages.BingoNumberDrawnNotify
typealias BingoStateNotify = Messages.BingoStateNotify
typealias BingoGameEndedNotify = Messages.BingoGameEndedNotify
typealias BingoRewardAnnouncedNotify = Messages.BingoRewardAnnouncedNotify
typealias BingoRewardAcquiredEvent = Messages.BingoRewardAcquiredEvent
typealias BingoRoomState = Messages.BingoRoomState
typealias BingoPlayerState = Messages.BingoPlayerState

val SubmitBingoCardReq.card: List<Int>
    get() = cardList

val BingoRoomState.drawnNumbers: List<Int>
    get() = drawnNumbersList

val BingoRoomState.players: List<BingoPlayerState>
    get() = playersList

val BingoRoomState.winners: List<String>
    get() = winnersList

val BingoPlayerState.card: List<Int>
    get() = cardList

val BingoPlayerState.marks: List<Boolean>
    get() = marksList

fun AuthenticateReq(accessToken: String): AuthenticateReq =
    Messages.AuthenticateReq.newBuilder()
        .setAccessToken(accessToken)
        .build()

fun AuthenticateRes(
    actorId: String,
    displayName: String,
    actorNodeRid: String,
): AuthenticateRes =
    Messages.AuthenticateRes.newBuilder()
        .setActorId(actorId)
        .setDisplayName(displayName)
        .setActorNodeRid(actorNodeRid)
        .build()

fun AuthenticatePlayerReq(accessToken: String): AuthenticatePlayerReq =
    Messages.AuthenticatePlayerReq.newBuilder()
        .setAccessToken(accessToken)
        .build()

fun AuthenticatePlayerRes(
    accepted: Boolean,
    actorId: String,
    displayName: String,
    reason: String?,
): AuthenticatePlayerRes =
    Messages.AuthenticatePlayerRes.newBuilder()
        .setAccepted(accepted)
        .setActorId(actorId)
        .setDisplayName(displayName)
        .apply {
            if (reason != null) {
                setReason(reason)
            }
        }
        .build()

fun EnsurePlayerActorReq(
    actorId: String,
    displayName: String,
    preferredActorNodeRid: String,
): EnsurePlayerActorReq =
    Messages.EnsurePlayerActorReq.newBuilder()
        .setActorId(actorId)
        .setDisplayName(displayName)
        .setPreferredActorNodeRid(preferredActorNodeRid)
        .build()

fun ActorRefWire(
    nodeRid: String,
    actorId: String,
    generation: Long,
): ActorRefWire =
    Messages.ActorRefWire.newBuilder()
        .setNodeRid(nodeRid)
        .setActorId(actorId)
        .setGeneration(generation)
        .build()

fun EnsurePlayerActorRes(
    actorId: String,
    actorType: String,
    actor: ActorRefWire,
): EnsurePlayerActorRes =
    Messages.EnsurePlayerActorRes.newBuilder()
        .setActorId(actorId)
        .setActorType(actorType)
        .setActor(actor)
        .build()

fun MatchBingoReq(mode: String): MatchBingoReq =
    Messages.MatchBingoReq.newBuilder()
        .setMode(mode)
        .build()

fun MatchBingoRes(
    roomId: String,
    state: BingoRoomState,
    roomOwnerNodeRid: String,
): MatchBingoRes =
    Messages.MatchBingoRes.newBuilder()
        .setRoomId(roomId)
        .setState(state)
        .setRoomOwnerNodeRid(roomOwnerNodeRid)
        .build()

fun MatchBingoApiReq(
    actorId: String,
    displayName: String,
    mode: String,
    actorNodeRid: String,
): MatchBingoApiReq =
    Messages.MatchBingoApiReq.newBuilder()
        .setActorId(actorId)
        .setDisplayName(displayName)
        .setMode(mode)
        .setActorNodeRid(actorNodeRid)
        .build()

fun MatchBingoApiRes(
    roomId: String,
    roomOwnerNodeRid: String,
): MatchBingoApiRes =
    Messages.MatchBingoApiRes.newBuilder()
        .setRoomId(roomId)
        .setRoomOwnerNodeRid(roomOwnerNodeRid)
        .build()

fun AllocateBingoRoomReq(
    actorId: String,
    mode: String,
    preferredOwnerNodeRid: String,
): AllocateBingoRoomReq =
    Messages.AllocateBingoRoomReq.newBuilder()
        .setActorId(actorId)
        .setMode(mode)
        .setPreferredOwnerNodeRid(preferredOwnerNodeRid)
        .build()

fun AllocateBingoRoomRes(
    roomId: String,
    roomOwnerNodeRid: String,
): AllocateBingoRoomRes =
    Messages.AllocateBingoRoomRes.newBuilder()
        .setRoomId(roomId)
        .setRoomOwnerNodeRid(roomOwnerNodeRid)
        .build()

fun BingoRoomSettingsPayload(
    mode: String,
    roomName: String,
    requiredPlayers: Int,
    maxDrawNumber: Int,
    drawPeriodMillis: Long,
    observedRoomId: String,
): BingoRoomSettingsPayload =
    Messages.BingoRoomSettingsPayload.newBuilder()
        .setMode(mode)
        .setRoomName(roomName)
        .setRequiredPlayers(requiredPlayers)
        .setMaxDrawNumber(maxDrawNumber)
        .setDrawPeriodMillis(drawPeriodMillis)
        .setObservedRoomId(observedRoomId)
        .build()

fun BingoRoomJoinReq(
    roomId: String,
    actorId: String,
    displayName: String,
    observeOnly: Boolean,
): BingoRoomJoinReq =
    Messages.BingoRoomJoinReq.newBuilder()
        .setRoomId(roomId)
        .setActorId(actorId)
        .setDisplayName(displayName)
        .setObserveOnly(observeOnly)
        .build()

fun BingoRoomJoinRes(state: BingoRoomState): BingoRoomJoinRes =
    Messages.BingoRoomJoinRes.newBuilder()
        .setState(state)
        .build()

fun SubmitBingoCardReq(
    roomId: String,
    card: List<Int>,
): SubmitBingoCardReq =
    Messages.SubmitBingoCardReq.newBuilder()
        .setRoomId(roomId)
        .addAllCard(card)
        .build()

fun SubmitBingoCardRes(state: BingoRoomState): SubmitBingoCardRes =
    Messages.SubmitBingoCardRes.newBuilder()
        .setState(state)
        .build()

fun ObserveBingoEventsReq(roomId: String): ObserveBingoEventsReq =
    Messages.ObserveBingoEventsReq.newBuilder()
        .setRoomId(roomId)
        .build()

fun ObserveBingoEventsRes(
    subscribed: Boolean,
    observerNodeRid: String,
): ObserveBingoEventsRes =
    Messages.ObserveBingoEventsRes.newBuilder()
        .setSubscribed(subscribed)
        .setObserverNodeRid(observerNodeRid)
        .build()

fun StopObservingBingoEventsReq(roomId: String): StopObservingBingoEventsReq =
    Messages.StopObservingBingoEventsReq.newBuilder()
        .setRoomId(roomId)
        .build()

fun StopObservingBingoEventsRes(
    stopped: Boolean,
    observerNodeRid: String,
): StopObservingBingoEventsRes =
    Messages.StopObservingBingoEventsRes.newBuilder()
        .setStopped(stopped)
        .setObserverNodeRid(observerNodeRid)
        .build()

fun PlayerJoinedNotify(
    roomId: String,
    actorId: String,
    displayName: String,
    seat: Int,
    isHost: Boolean,
    state: BingoRoomState,
): PlayerJoinedNotify =
    Messages.PlayerJoinedNotify.newBuilder()
        .setRoomId(roomId)
        .setActorId(actorId)
        .setDisplayName(displayName)
        .setSeat(seat)
        .setIsHost(isHost)
        .setState(state)
        .build()

fun BingoGameStartedNotify(state: BingoRoomState): BingoGameStartedNotify =
    Messages.BingoGameStartedNotify.newBuilder()
        .setState(state)
        .build()

fun BingoNumberDrawnNotify(
    roomId: String,
    drawSeq: Int,
    number: Int,
    state: BingoRoomState,
): BingoNumberDrawnNotify =
    Messages.BingoNumberDrawnNotify.newBuilder()
        .setRoomId(roomId)
        .setDrawSeq(drawSeq)
        .setNumber(number)
        .setState(state)
        .build()

fun BingoStateNotify(state: BingoRoomState): BingoStateNotify =
    Messages.BingoStateNotify.newBuilder()
        .setState(state)
        .build()

fun BingoGameEndedNotify(state: BingoRoomState): BingoGameEndedNotify =
    Messages.BingoGameEndedNotify.newBuilder()
        .setState(state)
        .build()

fun BingoRewardAnnouncedNotify(
    roomId: String,
    actorId: String,
    drawSeq: Int,
    itemId: String,
    itemName: String,
    rarity: String,
    receivingSpotNodeRid: String,
): BingoRewardAnnouncedNotify =
    Messages.BingoRewardAnnouncedNotify.newBuilder()
        .setRoomId(roomId)
        .setActorId(actorId)
        .setDrawSeq(drawSeq)
        .setItemId(itemId)
        .setItemName(itemName)
        .setRarity(rarity)
        .setReceivingSpotNodeRid(receivingSpotNodeRid)
        .build()

fun BingoRewardAcquiredEvent(
    roomId: String,
    actorId: String,
    drawSeq: Int,
    itemId: String,
    itemName: String,
    rarity: String,
): BingoRewardAcquiredEvent =
    Messages.BingoRewardAcquiredEvent.newBuilder()
        .setRoomId(roomId)
        .setActorId(actorId)
        .setDrawSeq(drawSeq)
        .setItemId(itemId)
        .setItemName(itemName)
        .setRarity(rarity)
        .build()

fun BingoRoomState(
    roomId: String,
    status: String,
    hostActorId: String,
    canStart: Boolean,
    drawSeq: Int,
    lastDrawnNumber: Int?,
    drawnNumbers: List<Int>,
    players: List<BingoPlayerState>,
    winners: List<String>,
): BingoRoomState =
    Messages.BingoRoomState.newBuilder()
        .setRoomId(roomId)
        .setStatus(status)
        .setHostActorId(hostActorId)
        .setCanStart(canStart)
        .setDrawSeq(drawSeq)
        .apply {
            if (lastDrawnNumber != null) {
                setLastDrawnNumber(lastDrawnNumber)
            }
        }
        .addAllDrawnNumbers(drawnNumbers)
        .addAllPlayers(players)
        .addAllWinners(winners)
        .build()

fun BingoPlayerState(
    actorId: String,
    displayName: String,
    seat: Int,
    isHost: Boolean,
    card: List<Int>,
    marks: List<Boolean>,
    completedLines: Int,
): BingoPlayerState =
    Messages.BingoPlayerState.newBuilder()
        .setActorId(actorId)
        .setDisplayName(displayName)
        .setSeat(seat)
        .setIsHost(isHost)
        .addAllCard(card)
        .addAllMarks(marks)
        .setCompletedLines(completedLines)
        .build()
